src/main/java/edu/cornell/kfs/sys/service/impl/AwsSecretServiceImpl.java

```diff
@@ -4,95 +4,169 @@
+        if (awsSecretsCache == null) {
+            LOG.debug("createCacheIfNotPresent, Cache has not been setup, so run the clear cache to create a new cache element");
+            clearCache();
+        }
+    }
+    
     @Override
     public String getSingleStringValueFromAwsSecret(String awsKeyName, boolean useKfsInstanceNamespace) {
+        createCacheIfNotPresent();
         String fullAwsKey = buildFullAwsKeyName(awsKeyName, useKfsInstanceNamespace);
-        GetSecretValueRequest getSecretValueRequest = new GetSecretValueRequest().withSecretId(fullAwsKey);
-        GetSecretValueResult getSecretValueResult = null;
+        String cachedSecretValue = retrieveSecretFromCache(fullAwsKey);
+        
+        if (StringUtils.isNotBlank(cachedSecretValue)) {
+            return cachedSecretValue;
+        } else {
+            LOG.debug("getSingleStringValueFromAwsSecret, in cache there was no value for " + fullAwsKey);
+            String secretValue = retrieveSecretFromAws(fullAwsKey);
+            updateCacheValue(fullAwsKey, secretValue);
+            return secretValue;
+        }
+    }
+
+    protected String retrieveSecretFromCache(String fullAwsKey) {
+        if (fullAwsKey == null) {
+            throw new RuntimeException(A_NULL_AWS_KEY_IS_NOT_ALLOWED);
+        }
+        if (awsSecretsCache.containsKey(fullAwsKey)) {
+            LOG.debug("retrieveSecretFromCache, in cache found a secret value for " + fullAwsKey);
+            return awsSecretsCache.get(fullAwsKey);
+        } else {
+            return null;
+        }
+    }
+    
+    private void updateCacheValue(String fullAwsKey, String secretValue) {
+        if (fullAwsKey == null) {
+            LOG.debug("updateCacheValue, the AWS key provided was null, so can't cache it.");
+        } else if (secretValue == null) {
+            LOG.debug("updateCacheValue, a null value was provided for AWS key " + fullAwsKey + " so it can not be cached");
+        } else {
+            LOG.debug("updateCacheValue, in cache setting value for secret key " + fullAwsKey);
+            awsSecretsCache.put(fullAwsKey, secretValue);
+        }
+    }
+    
+    protected String retrieveSecretFromAws(String fullAwsKey) {
+        SecretsManagerClient client = buildSecretsManagerClient();
         
-        AWSSecretsManager client = buildAWSSecretsManager();
         try {
-            getSecretValueResult = client.getSecretValue(getSecretValueRequest);
-        } catch (SdkClientException e) {
+            GetSecretValueRequest valueRequest = GetSecretValueRequest.builder()
+                    .secretId(fullAwsKey)
+                    .build();
+            GetSecretValueResponse valueResponse = client.getSecretValue(valueRequest);
+            return valueResponse.secretString();
+        } catch (SecretsManagerException e) {
             LOG.error("getSingleStringValueFromAwsSecret, had an error getting value for secret " + fullAwsKey, e);
```

Comment:
I think this log message should be updated to reference the "retrieveSecretFromAws" method instead.

Reply:
good catch